website/docs/src/content/cloud/storage/search.mdx

```diff
@@ -0,0 +1,487 @@
+      results.map((r) => r.attributes?.text),
+      null,
+      2,
+    );
+  },
+});
+```
+
+### Step 3: Create the RAG agent
+
+Now, create an agent that uses the query tool to access relevant information:
+
+```tsx
+const RagAgent = gensx.Component("RagAgent", ({ question }) => (
+  <GSXChatCompletion
+    messages={[
+      {
+        role: "system",
+        content: `You are a baseball expert assistant. Use the query tool to
+          look up relevant information before answering questions.`,
+      },
+      { role: "user", content: question },
+    ]}
+    model="gpt-4o-mini"
+    tools={[queryTool]}
+  >
+    {(result) => result.choices[0].message.content}
+  </GSXChatCompletion>
+));
+```
+
+### Step 4: Combine Everything in a Workflow
+
+Finally, put it all together in a workflow:
+
+```tsx
+const RagWorkflow = gensx.Component(
+  "RagWorkflow",
+  async ({ question, shouldReindex }) => {
+    // Optionally reindex documents
+    if (shouldReindex) {
+      await PrepareDocuments();
+    }
+
+    // Use the RAG agent to answer the question
+    return <RagAgent question={question} />;
+  },
+);
+```
+
+## Practical examples
+
+### Agent memory system
+
+One powerful application of vector search is creating a long-term memory system for AI agents:
+
+```tsx
+import * as gensx from "@gensx/core";
+import { OpenAIEmbedding } from "@gensx/openai";
+import { useSearch } from "@gensx/storage";
+
+// Component to store a memory
+const StoreMemory = gensx.Component(
+  "StoreMemory",
+  async ({ userId, memory, importance = "medium" }) => {
+    const search = await useSearch(`memories-${userId}`);
+
+    // Generate embedding for this memory
+    const embedding = await OpenAIEmbedding.run({
+      model: "text-embedding-3-small",
+      input: memory,
+    });
+
+    // Store the memory with metadata
+    await search.upsert({
+      vectors: [
+        {
+          id: `memory-${Date.now()}`,
+          vector: embedding.data[0].embedding,
+          attributes: {
+            content: memory,
+            timestamp: new Date().toISOString(),
+            importance: importance, // "high", "medium", "low"
+            source: "user-interaction",
+          },
+        },
+      ],
+      distanceMetric: "cosine_distance",
+    });
+
+    return { success: true };
+  },
+);
+
+// Component to recall relevant memories
+const RecallMemories = gensx.Component(
+  "RecallMemories",
+  async ({ userId, context, maxResults = 5 }) => {
+    const search = await useSearch(`memories-${userId}`);
+
+    // Generate embedding for the context
+    const embedding = await OpenAIEmbedding.run({
+      model: "text-embedding-3-small",
+      input: context,
+    });
+
+    // Query for relevant memories, prioritizing important ones
+    const results = await search.query({
+      vector: embedding.data[0].embedding,
+      topK: maxResults,
+      includeAttributes: true,
+      // Optional: rank by both relevance and importance
+      rankBy: {
```

Comment:
i don't think rankBy works like this. I noticed it's also incorrect other places.

```
export type RankBy_OrderByAttribute = [string, "asc" | "desc"];
export type RankBy_Text = [string, "BM25", string | [string]] | ["Sum", RankBy_Text[]] | ["Product", [RankBy_Text, number]] | ["Product", [number, RankBy_Text]];
export type RankBy = RankBy_Text | RankBy_OrderByAttribute;
```